File: src/utils/gradientColormap.ts
import { AnalysisOptions, GradientStop } from '../types/dataset';

export const defaultStops = (): GradientStop[] => [
  { position: 0, color: '#0000ff' }, { position: 1, color: '#ff0000' },
];
export const defaultAnalysis = (): AnalysisOptions => ({
  mode: 'none', exact_temperature: 30, tolerance: 0.1,
  range_min: 30, range_max: 35, flat_color: '#ff4500', stops: null,
});
export const gradientCss = (stops: GradientStop[]) =>
  `linear-gradient(90deg, ${[...stops].sort((a, b) => a.position - b.position)
    .map(s => `${s.color} ${s.position * 100}%`).join(', ')})`;

export function interpolateColor(stops: GradientStop[], position: number): string {
  const sorted = [...stops].sort((a, b) => a.position - b.position);
  const right = sorted.find(s => s.position >= position) ?? sorted[sorted.length - 1];
  const left = [...sorted].reverse().find(s => s.position <= position) ?? sorted[0];
  const ratio = right.position === left.position ? 0 : (position - left.position) / (right.position - left.position);
  return '#' + [1, 3, 5].map(i => Math.round(parseInt(left.color.slice(i, i + 2), 16) * (1 - ratio)
    + parseInt(right.color.slice(i, i + 2), 16) * ratio).toString(16).padStart(2, '0')).join('');
}

export function analysisError(value: AnalysisOptions): string | null {
  if (![value.exact_temperature, value.tolerance, value.range_min, value.range_max].every(Number.isFinite)) return 'Enter finite numeric values.';
  if (value.tolerance < 0) return 'Tolerance must be non-negative.';
  if (value.range_min > value.range_max) return 'Range minimum must be less than or equal to maximum.';
  return null;
}

export function analysisUnits(units?: string): string {
  const key = (units ?? '').toLowerCase().replace(/[ _]/g, '');
  return ['k', 'kelvin', 'degreekelvin', 'degreeskelvin', 'f', '°f', 'degf', 'degreefahrenheit', 'degreesfahrenheit'].includes(key) ? '°C' : (units || 'dataset units');
}

File: server_backup_before_argo/pipeline/gradient_colormap.py
"""Provider-independent scalar analysis and continuous RGB gradients."""
from typing import Literal, Optional
import numpy as np
from pydantic import BaseModel, ConfigDict, Field, model_validator


class GradientStop(BaseModel):
    model_config = ConfigDict(extra="forbid", allow_inf_nan=False)
    position: float = Field(ge=0, le=1)
    color: str = Field(pattern=r"^#[0-9a-fA-F]{6}$")


class AnalysisOptions(BaseModel):
    model_config = ConfigDict(extra="forbid", allow_inf_nan=False)
    mode: Literal["none", "exact", "range"] = "none"
    exact_temperature: float = 30
    tolerance: float = Field(default=0.1, ge=0)
    range_min: float = 30
    range_max: float = 35
    flat_color: str = Field(default="#ff4500", pattern=r"^#[0-9a-fA-F]{6}$")
    stops: Optional[list[GradientStop]] = Field(default=None, min_length=2, max_length=32)

    @model_validator(mode="after")
    def validate_range_and_stops(self):
        if self.range_min > self.range_max:
            raise ValueError("Range minimum must be <= maximum")
        if self.stops:
            self.stops = sorted(self.stops, key=lambda stop: stop.position)
            if any(a.position == b.position for a, b in zip(self.stops, self.stops[1:])):
                raise ValueError("Gradient stop positions must be distinct")
        return self


def rgb(color):
    return np.array([int(color[i:i + 2], 16) / 255 for i in (1, 3, 5)])


def gradient_rgba(normalized, stops):
    stops = stops or [GradientStop(position=0, color="#0000ff"), GradientStop(position=1, color="#ff0000")]
    positions = [s.position for s in stops]
    colors = np.array([rgb(s.color) for s in stops])
    rgba = np.ones((*normalized.shape, 4), dtype=float)
    for channel in range(3):
        rgba[..., channel] = np.interp(normalized, positions, colors[:, channel])
    return rgba


def analysis_values(frame, options):
    """Filtering uses Celsius for known temperature units; points retain native units."""
    values = np.asarray(frame.data, dtype=float)
    units = (frame.units or "").lower().replace(" ", "").replace("_", "")
    if options.mode != "none":
        if units in {"k", "kelvin", "degreekelvin", "degreeskelvin"}:
            return values - 273.15, "°C"
        if units in {"f", "°f", "degf", "degreefahrenheit", "degreesfahrenheit"}:
            return (values - 32) * 5 / 9, "°C"
    return values, frame.units


def analyze_slice(frame, options=None):
    options = options or AnalysisOptions()
    values, units = analysis_values(frame, options)
    valid = np.isfinite(frame.data)
    if frame.mask is not None:
        valid &= ~frame.mask
    if frame.fill_value is not None:
        valid &= frame.data != frame.fill_value
    matching = valid.copy()
    if options.mode == "exact":
        matching &= np.abs(values - options.exact_temperature) <= options.tolerance
    elif options.mode == "range":
        matching &= (values >= options.range_min) & (values <= options.range_max)
    selected = values[matching]
    count, total = int(matching.sum()), int(valid.sum())
    stats = {
        "variable": frame.variable_name, "units": units,
        "timestamp": frame.timestamp, "time_index": frame.time_index,
        "matching_cells": count, "valid_cells": total,
        "total_cells": int(values.size),
        "matching_percent": 100 * count / total if total else 0,
        "min": float(selected.min()) if count else None,
        "max": float(selected.max()) if count else None,
        "mean": float(selected.mean()) if count else None,
        "message": None if count else "No dataset values found for the selected filter and frame.",
    }
    return values, matching, stats
